src/commonMain/kotlin/io/konform/validation/ValidationBuilder.kt

```diff
@@ -1,70 +1,111 @@
 package io.konform.validation
 
-import io.konform.validation.internal.ArrayValidation
-import io.konform.validation.internal.IterableValidation
-import io.konform.validation.internal.MapValidation
-import io.konform.validation.internal.OptionalValidation
-import io.konform.validation.internal.RequiredValidation
-import io.konform.validation.internal.ValidationBuilderImpl
+import io.konform.validation.internal.*
 import kotlin.jvm.JvmName
+import kotlin.reflect.KFunction1
 import kotlin.reflect.KProperty1
 
 @DslMarker
 private annotation class ValidationScope
 
 @ValidationScope
-abstract class ValidationBuilder<T> {
-    abstract fun build(): Validation<T>
-    abstract fun addConstraint(errorMessage: String, vararg templateValues: String, test: (T) -> Boolean): Constraint<T>
-    abstract infix fun Constraint<T>.hint(hint: String): Constraint<T>
-    abstract operator fun <R> KProperty1<T, R>.invoke(init: ValidationBuilder<R>.() -> Unit)
-    internal abstract fun <R> onEachIterable(prop: KProperty1<T, Iterable<R>>, init: ValidationBuilder<R>.() -> Unit)
+abstract class ValidationBuilder<C, T, E> : ComposableBuilder<C, T, E> {
+    abstract override fun build(): Validation<C, T, E>
+
+    abstract fun addConstraint(hint: HintBuilder<C, T, E>, vararg values: Any, test: C.(T) -> Boolean): ConstraintBuilder<C, T, E>
+
+    abstract operator fun <R> KProperty1<T, R>.invoke(init: ValidationBuilder<C, R, E>.() -> Unit)
+    abstract operator fun <R> KFunction1<T, R>.invoke(init: ValidationBuilder<C, R, E>.() -> Unit)
+
+    internal abstract fun <R> onEachIterable(name: String, mapFn: (T) -> Iterable<R>, init: ValidationBuilder<C, R, E>.() -> Unit)
     @JvmName("onEachIterable")
-    infix fun <R> KProperty1<T, Iterable<R>>.onEach(init: ValidationBuilder<R>.() -> Unit) = onEachIterable(this, init)
-    internal abstract fun <R> onEachArray(prop: KProperty1<T, Array<R>>, init: ValidationBuilder<R>.() -> Unit)
+    infix fun <R> KProperty1<T, Iterable<R>>.onEach(init: ValidationBuilder<C, R, E>.() -> Unit) = onEachIterable(this.name, this, init)
+    @JvmName("onEachIterable")
+    infix fun <R> KFunction1<T, Iterable<R>>.onEach(init: ValidationBuilder<C, R, E>.() -> Unit) = onEachIterable(this.name, this, init)
```

Comment:
This and the other extension functions on `KFunction1` are not tested as far as I can see.

Reply:
You are right. I added some tests in a jvmTest module.